### PR TITLE
feat(wallet): add SafeSign to LiFi bridge transactions

### PR DESCRIPTION
**Pull request by onyb**
Resolves https://github.com/brave/brave-browser/issues/39701

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

See STR in linked issue. Note that this PR does not implement UI for SafeSign in bridge transactions, which will be implemented in a subsequent PR.

### CODE REVIEW

File: components/brave_wallet_ui/utils/tx-utils.ts [hunk 1]
```diff
@@ -422,6 +422,11 @@ export const findTransactionToken = <
   )
 }
 
+// Use this function to synchronously extract swap/bridge details from a
+// transaction.
+//
+// Prefer using useSwapTransactionParser() hook in React components, which
+// can asynchronously extract details from the blockchain.
```
Comment (josheleonard) [inline components/brave_wallet_ui/utils/tx-utils.ts:429]: Mind using `/** */` style comments here so the notes appear when the function is hovered-over in a text editor?

**Review by josheleonard — APPROVED** [review]
Frontend LGTM. Great work!

**Review by supermassive — APPROVED** [review]
wallet core lgtm

**Review by nuo-xu — APPROVED** [review]
fixed the iOS side. @StephenHeaps PTAL on my changes.

File: ios/brave-ios/Sources/BraveWallet/Crypto/Stores/SwapTokenStore.swift [hunk 2]
```diff
@@ -376,7 +376,9 @@ public class SwapTokenStore: ObservableObject, WalletObserverStore {
       toAmount: buyAmountInWei,
       slippagePercentage: "\(slippagePercentage)",
       routePriority: .recommended,
-      // TODO(stephenheaps): Enable Lifi support https://github.com/brave/brave-browser/issues/36436
+      // TODO(stephenheaps): Enable Lifi support
+      // https://github.com/brave/brave-browser/issues/36436
```
Comment (StephenHeaps) [inline ios/brave-ios/Sources/BraveWallet/Crypto/Stores/SwapTokenStore.swift:380]: Should remove this old url, just link to the new issue #39043

File: ios/brave-ios/Sources/BraveWallet/Crypto/Stores/AccountActivityStore.swift [hunk 3]
```diff
@@ -430,6 +435,7 @@ class AccountActivityStore: ObservableObject, WalletObserverStore {
   private func buildTransactionSections(
     transactions: [BraveWallet.TransactionInfo],
     networksForCoin: [BraveWallet.CoinType: [BraveWallet.NetworkInfo]],
+    allNetworks: [BraveWallet.NetworkInfo],
```
Comment (StephenHeaps) [inline ios/brave-ios/Sources/BraveWallet/Crypto/Stores/AccountActivityStore.swift:438]: Should just remove `networksForCoin` if we're passing in all networks now.

File: ios/brave-ios/Sources/BraveWallet/Crypto/Transactions/TransactionParser.swift [hunk 4]
```diff
@@ -126,7 +126,8 @@ enum TransactionParser {
 
   static func parseTransaction(
     transaction: BraveWallet.TransactionInfo,
-    network: BraveWallet.NetworkInfo,
+    currentNetwork: BraveWallet.NetworkInfo,
```
Comment (StephenHeaps) [inline ios/brave-ios/Sources/BraveWallet/Crypto/Transactions/TransactionParser.swift:129]: `currentNetwork` isn't needed anymore, can fetch it using the `chainId` & `coin` from the `TransactionInfo` model.

File: ios/brave-ios/Sources/BraveWallet/Crypto/Transactions/TransactionParser.swift [hunk 5]
```diff
@@ -138,41 +139,41 @@ enum TransactionParser {
   ) -> ParsedTransaction? {
     let fromAccountInfo =
       accountInfos.first(where: { $0.accountId == transaction.fromAccountId }) ?? .init()
-    let formatter = WalletAmountFormatter(
-      decimalFormatStyle: decimalFormatStyle ?? .decimals(precision: Int(network.decimals))
-    )
```
Comment (StephenHeaps) [inline ios/brave-ios/Sources/BraveWallet/Crypto/Transactions/TransactionParser.swift:143]: Why did this move to be defined in each `TransactionType`? Seems unnecessary code duplication, we're still using the same initialization in each type?

File: ios/brave-ios/Sources/BraveWallet/Crypto/Transactions/TransactionParser.swift [hunk 6]
```diff
@@ -1106,6 +1150,8 @@ struct SendDetails: Equatable {
 struct EthSwapDetails: Equatable {
   /// Token being swapped from
   let fromToken: BraveWallet.BlockchainToken?
+  /// From token network (not yet being used. preparation for bridging)
+  let fromNetwork: BraveWallet.NetworkInfo
```
Comment (StephenHeaps) [inline ios/brave-ios/Sources/BraveWallet/Crypto/Transactions/TransactionParser.swift:1154]: nit unneeded comment
```suggestion
  /// From token network
  let fromNetwork: BraveWallet.NetworkInfo
```

File: ios/brave-ios/Sources/BraveWallet/Crypto/Transactions/TransactionParser.swift [hunk 7]
```diff
@@ -1115,6 +1161,8 @@ struct EthSwapDetails: Equatable {
 
   /// Token being swapped to
   let toToken: BraveWallet.BlockchainToken?
+  /// To token network (not yet being used. preparation for bridging)
+  let toNetwork: BraveWallet.NetworkInfo
```
Comment (StephenHeaps) [inline ios/brave-ios/Sources/BraveWallet/Crypto/Transactions/TransactionParser.swift:1165]: nit: unneeded comment
```suggestion
  /// To token network (not yet being used. preparation for bridging)
  let toNetwork: BraveWallet.NetworkInfo
```

**Review by StephenHeaps — CHANGES_REQUESTED** [review]

File: ios/brave-ios/Sources/BraveWallet/Crypto/Stores/TransactionDetailsStore.swift [hunk 8]
```diff
@@ -117,8 +117,8 @@ class TransactionDetailsStore: ObservableObject, WalletObserverStore {
 
   func update() {
     Task { @MainActor in
-      let coin = transaction.coin
-      let networksForCoin = await rpcService.allNetworks().filter({ $0.coin == coin })
+      let allNetworks = await rpcService.allNetworks()
+      let networksForCoin = allNetworks.filter({ $0.coin == transaction.coin })
       guard let network = networksForCoin.first(where: { $0.chainId == transaction.chainId }) else {
```
Comment (StephenHeaps) [inline ios/brave-ios/Sources/BraveWallet/Crypto/Stores/TransactionDetailsStore.swift:122]: Just check `allNetworks` here now so we only loop through once

File: ios/brave-ios/Sources/BraveWallet/Crypto/Transactions/TransactionParser.swift [hunk 9]
```diff
@@ -714,6 +741,9 @@ enum TransactionParser {
         )
       )
     case .solanaDappSignAndSendTransaction, .solanaDappSignTransaction, .solanaSwap:
+      let formatter = WalletAmountFormatter(
+        decimalFormatStyle: decimalFormatStyle ?? .decimals(precision: Int(txNetwork.decimals))
+      )
```
Comment (StephenHeaps) [inline ios/brave-ios/Sources/BraveWallet/Crypto/Transactions/TransactionParser.swift:746]: Shouldn't be needed here

**Review by StephenHeaps — APPROVED** [review]

**Comment by srirambv** [comment]
Verification passed on 

Brave | 1.70.5 Chromium: 127.0.6533.57 (Official Build) nightly (64-bit)
-- | --
Revision | face8d58ab81422480c8c05b9ba5d518e1a2d227
OS | Windows 11 Version 23H2 (Build 22631.3880)

- Verified steps from issue description
- Verified Safe Sign panel is shown for bridge transaction

https://github.com/user-attachments/assets/c48c5644-8be4-40fd-9742-dc7f89e6c6d1

**Pull request merged**